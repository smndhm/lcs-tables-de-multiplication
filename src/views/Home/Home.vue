<template>
  <header>
    <h1>{{ $t("multiplication") }}</h1>
  </header>
  <main>
    <fieldset>
      <legend v-html="$t('home.legend')"></legend>
      <ul>
        <li v-for="table in tables" :key="table">
          <label :for="`table-${table}`"
            ><input
              type="checkbox"
              v-model="selectedTables"
              :value="table"
              :id="`table-${table}`"
              name="table"
              @change="updateSelectedTables"
            />
            <span>{{ table }}</span></label
          >
        </li>
        <li>
          <label for="table-all"
            ><input
              type="checkbox"
              id="table-all"
              ref="table-all"
              @change="allSelected"
            />
            <span>{{ $t("home.all") }}</span></label
          >
        </li>
      </ul>
    </fieldset>
  </main>
  <footer>
    <router-link
      class="btn"
      :class="{ 'btn--disabled': !selectedTables.length }"
      :to="{
        name: 'Multiplications',
        query: {
          tables: this.selectedTables,
        },
      }"
    >
      {{ $t("home.start") }}
    </router-link>
  </footer>
</template>

<script>
export default {
  name: "Home",
  data() {
    return {
      tables: [2, 3, 4, 5, 6, 7, 8, 9],
      selectedTables: [],
    };
  },
  mounted() {
    if (localStorage.getItem("selectedTables")) {
      this.selectedTables = JSON.parse(localStorage.getItem("selectedTables"));
    }
    this.updateSelectedTables();
  },
  methods: {
    storeSelectedTables() {
      localStorage.setItem(
        "selectedTables",
        JSON.stringify(this.selectedTables)
      );
    },
    allSelected() {
      if (this.$refs["table-all"].checked) {
        this.selectedTables = this.tables;
      } else {
        this.selectedTables = [];
      }
      this.storeSelectedTables();
    },
    updateSelectedTables() {
      this.$refs["table-all"].checked = false;
      this.$refs["table-all"].indeterminate = false;
      if (this.selectedTables.length === this.tables.length) {
        this.$refs["table-all"].checked = true;
      } else if (
        this.selectedTables.length < this.tables.length &&
        this.selectedTables.length !== 0
      ) {
        this.$refs["table-all"].indeterminate = true;
      }
      this.storeSelectedTables();
    },
  },
};
</script>

<style lang="scss" scoped>
h1 {
  font-size: calc(var(--font-size) * 2);
}

fieldset {
  border: none;
  legend {
    font-size: calc(var(--font-size) * 1.125);
  }
}

ul {
  // Display
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  gap: 1rem;
  padding: 0;
  // Decoration
  list-style: none;

  li {
    label {
      input[type="checkbox"] {
        position: absolute;
        opacity: 0;
      }

      span {
        font-size: calc(var(--font-size) * 1.125);
        display: flex;
        justify-content: center;
        align-items: center;
        border: 0.2rem solid;
        border-radius: 1em;
        padding: 0 0.5em;
        min-width: 2em;
        height: 2em;
      }

      &:hover span,
      input[type="checkbox"]:focus + span {
        cursor: pointer;
        box-shadow: 0 0 0 0.4rem rgba(0, 0, 0, 0.1);
      }

      input[type="checkbox"]:active + span {
        box-shadow: 0 0 0 0.4rem rgba(0, 0, 0, 0.2);
      }

      input[type="checkbox"]:checked:not(:indeterminate) + span {
        box-shadow: 0 0 0 0.2rem rgba(0, 0, 0, 0.6);
      }

      input[type="checkbox"]:checked:focus + span,
      &:hover input[type="checkbox"]:checked + span {
        box-shadow: 0 0 0 0.4rem rgba(0, 0, 0, 0.6);
      }

      input[type="checkbox"]:checked:focus:active + span {
        box-shadow: 0 0 0 0.4rem rgba(0, 0, 0, 0.8);
      }
    }
  }
}
</style>
