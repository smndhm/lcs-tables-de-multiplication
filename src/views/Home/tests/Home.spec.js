import { config, mount } from "@vue/test-utils";
const { axe, toHaveNoViolations } = require("jest-axe");

expect.extend(toHaveNoViolations);

import Home from "../Home";

config.global.mocks = {
  $t: (key) => key,
};

describe("Home.vue", () => {
  test("Should renders view", async () => {
    const wrapper = mount(Home);
    const html = wrapper.html();
    expect(html).toMatchSnapshot();
    expect(await axe(html)).toHaveNoViolations();
  });

  test("Should manage selected tables", async () => {
    const wrapper = mount(Home);
    //Select one
    wrapper.get("#table-2").setChecked(true);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedTables).toStrictEqual([2]);
    // Select all
    wrapper.get({ ref: "table-all" }).setChecked(true);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedTables).toStrictEqual([2, 3, 4, 5, 6, 7, 8, 9]);
    // Unselect all
    wrapper.get({ ref: "table-all" }).setChecked(false);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedTables).toStrictEqual([]);
    // Select all one by one
    for (let i = 2; i <= 9; i++) {
      wrapper.get(`#table-${i}`).setChecked(true);
      await wrapper.vm.$nextTick();
    }
    expect(wrapper.vm.selectedTables).toStrictEqual([2, 3, 4, 5, 6, 7, 8, 9]);
  });
});
