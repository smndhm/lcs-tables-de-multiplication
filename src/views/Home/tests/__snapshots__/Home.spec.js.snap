// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue Should renders view 1`] = `
<header>
  <h1>multiplication</h1>
</header>
<main>
  <fieldset>
    <legend>home.legend</legend>
    <ul>
      <li><label for="table-2"><input type="checkbox" id="table-2" name="table" value="2"><span>2</span></label></li>
      <li><label for="table-3"><input type="checkbox" id="table-3" name="table" value="3"><span>3</span></label></li>
      <li><label for="table-4"><input type="checkbox" id="table-4" name="table" value="4"><span>4</span></label></li>
      <li><label for="table-5"><input type="checkbox" id="table-5" name="table" value="5"><span>5</span></label></li>
      <li><label for="table-6"><input type="checkbox" id="table-6" name="table" value="6"><span>6</span></label></li>
      <li><label for="table-7"><input type="checkbox" id="table-7" name="table" value="7"><span>7</span></label></li>
      <li><label for="table-8"><input type="checkbox" id="table-8" name="table" value="8"><span>8</span></label></li>
      <li><label for="table-9"><input type="checkbox" id="table-9" name="table" value="9"><span>9</span></label></li>
      <li><label for="table-all"><input type="checkbox" id="table-all"><span>home.all</span></label></li>
    </ul>
  </fieldset>
</main>
<footer>
  <router-link class="btn btn--disabled" to="[object Object]">home.start</router-link>
</footer>
`;
